paymentsheet/src/main/java/com/stripe/android/paymentsheet/elements/common/TextFieldElement.kt

```diff
@@ -0,0 +1,108 @@
+package com.stripe.android.paymentsheet.elements.common
+
+import android.util.Log
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.Transformations
+import androidx.lifecycle.distinctUntilChanged
+import com.stripe.android.paymentsheet.NotNullMutableLiveData
+import com.stripe.android.paymentsheet.R
+
+/**
+ * This class will provide the onValueChanged and onFocusChanged functionality to the element's
+ * composable.  These functions will update the observables as needed.  It is responsible for
+ * exposing immutable observers for its data
+ */
+internal open class TextFieldElement(private val textFieldConfig: TextFieldConfigInterface) {
+    val debugLabel = textFieldConfig.debugLabel
+    private val isDebug = true
+
+    /** This is all the information that can be observed on the element */
+    private val _input: NotNullMutableLiveData<String> = NotNullMutableLiveData("")
+    val input: LiveData<String> = _input.distinctUntilChanged()
+
+    private val _elementState: NotNullMutableLiveData<TextFieldElementState> =
+        NotNullMutableLiveData(
+            TextFieldElementState.TextFieldElementStateError(
+                R.string.invalid
+            )
+        )
+
+    private val _hasFocus: NotNullMutableLiveData<Boolean> = NotNullMutableLiveData(false)
+
+    private val _visibleError = MediatorLiveData<Boolean>().apply {
+        addSource(_elementState) { postValue(shouldShowError(it, _hasFocus.value)) }
+        addSource(_hasFocus) { postValue(shouldShowError(_elementState.value, it)) }
+        postValue(false)
+    }
+    val visibleError: LiveData<Boolean> = _visibleError.distinctUntilChanged()
+    val errorMessage: LiveData<Int> = Transformations.map(_visibleError) { visibleError ->
+        _elementState.value.getErrorMessageResId()?.takeIf { visibleError }
+    }
+    val isFull = Transformations.map(_elementState) { it.isFull() }
+    val isComplete = Transformations.map(_elementState) { it.isValid() }
```

Comment:
Just wondering if you have considered using `StateFlow`/`MutableStateFlow` instead of `LiveData`. It makes things simpler because it's all kotlin (nullability in `LiveData` is bad because it's all implemented in Java) and it also has a much better operator ecosystem backed in. There are the `LiveData` ktx extensions but that only does so much and, in the end, everything needs to go into a `MediatorLiveData`.

Reply:
I will post a review for this when this one is complete.